### PR TITLE
Remove ForEach functions from arrays

### PR DESCRIPTION
**Pull request by Raelr**
### Description

This PR removes all usages of the `ForEach`, `MForEach`, `ForEachI`, and `MForEachI` functions from the custom array functions.

### The PR has been...
- [x] provided a reasonable name that is not just the branch name (e.g "Added Vulkan render delegate")
- [x] linked to its related issue
- [x] assigned a reviewer from the team
- [x] labelled appropriately

### The code has been...
- [x] made mergable and free of conflicts in relation to `master` *(according to GitHub)*
- [x] tested in a packaged state using the `package` targets
- [ ] pulled to the reviewer's machine and reasonably tested

### CODE REVIEW

File: engine/render/renderer/platform/vulkan/Pipeline.cpp [hunk 1]
```diff
@@ -99,24 +99,30 @@ Pipeline Pipeline::Builder::Build()
...
+        for (auto attrIt = attributes.CreateIterator(); attrIt; ++attrIt)
+        {
+            auto& vertex = *attrIt;
+            auto attributeIndex = attrIt.GetIndex() + processedAttributes;
 
-            attributeDescriptions[attributeIndex] = {static_cast<uint32_t>(j),
-                                                     static_cast<uint32_t>(i),
+            attributeDescriptions[attributeIndex] = {static_cast<uint32_t>(attrIt.GetIndex()),
+                                                     static_cast<uint32_t>(vertIt.GetIndex()),
                                                      Utils::ToVkFormat(vertex.type),
```
Comment (J-Mo63) [inline engine/render/renderer/platform/vulkan/Pipeline.cpp:120]: Looks like these can be `const`? Also, you can avoid caching `index`, if you just plan to call the function again when you need it anyway.

File: engine/render/renderer/platform/vulkan/Material.cpp [hunk 2]
```diff
@@ -151,26 +163,34 @@ uint32_t Material::SetTexture(Hash::StringId id, Texture2D* texture)
...
-    propertiesSlots.MForEachI([&](PropertiesSlot& slot, size_t i) {
+    for (auto it = propertiesSlots.CreateIterator(); it; ++it)
+    {
+        auto& slot = *it;
         auto& properties = slot.properties;
+
         auto propIdx = FindPropertyIndex(id, slot);
 
-        if (propIdx == -1) return;
+        if (propIdx == -1) continue;
```
Comment (J-Mo63) [inline engine/render/renderer/platform/vulkan/Material.cpp:173]: Great catch on this

File: engine/render/renderer/platform/vulkan/Material.cpp [hunk 3]
```diff
@@ -337,26 +371,28 @@ void Material::QueuePropertyUpdate(MSArray<VkWriteDescriptorSet, 10>& writeQueue
 int32_t Material::FindPropertyIndex(Hash::StringId id, PropertiesSlot& slot)
 {
     int32_t foundIdx {-1};
-    slot.properties.MForEachI([&](Property& property, size_t i) {
-        if (property.id == id)
+    for (auto it = slot.properties.CreateIterator(); it; ++it)
+    {
+        if ((*it).id == id)
```
Comment (J-Mo63) [inline engine/render/renderer/platform/vulkan/Material.cpp:376]: `if (it->id == id)`?

**Review by J-Mo63 — APPROVED** [review]
Thanks for this one, it's now much easier to see what's happening across the material class.

File: engine/render/renderer/platform/vulkan/Pipeline.cpp (hunk 1 above)
Comment (Raelr) [reply engine/render/renderer/platform/vulkan/Pipeline.cpp:120]: Changed.

File: engine/render/renderer/platform/vulkan/Material.cpp (hunk 3 above)
Comment (Raelr) [reply engine/render/renderer/platform/vulkan/Material.cpp:376]: Added the -> operator to the iterators to enable this.

**Pull request merged**